test/framework/event-wrapper.ts

```diff
@@ -0,0 +1,329 @@
+import {expect} from 'chai'
+import {
+    Contract,
+    ContractReceipt,
+    ContractTransaction,
+    Event,
+    EventFilter,
+    utils
+} from 'ethers'
+import {TypedEvent, TypedEventFilter} from '../../typechain-types/common'
+import {EventListener} from './event-listener'
+
+function eventsArg(
+    name: string,
+    receipt: ContractReceipt,
+    emitter?: Contract
+): unknown[] {
+    const found: unknown[] = []
+
+    // eslint-disable-next-line no-undefined
+    if (emitter === undefined) {
+        for (const entry of receipt.events ?? []) {
+            if (entry?.event === name) {
+                // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+                found.push(entry.args!)
+            }
+        }
+    } else {
+        const addr = emitter.address.toUpperCase()
+        const parser = emitter.interface
+
+        const fragment = parser.getEvent(name)
+        const id = utils.id(fragment.format())
+
+        for (const entry of receipt.logs) {
+            if (
+                entry.topics[0] === id &&
+                entry.address.toUpperCase() === addr
+            ) {
+                const parsed = parser.decodeEventLog(
+                    fragment,
+                    entry.data,
+                    entry.topics
+                )
+                found.push(parsed)
+            }
+        }
+    }
+
+    expect(
+        found.length,
+        `Failed to find any event matching name: ${name}`
+    ).is.greaterThan(0)
+
+    return found
+}
+
+export type ContractReceiptSource =
+    | ContractReceipt
+    | Promise<ContractReceipt>
+    | ContractTransaction
+    | Promise<ContractTransaction>
+
+export interface EventFactory<TArgObj = unknown> {
+    one<Result = TArgObj>(
+        receipt: ContractReceipt,
+        fn?: (args: TArgObj) => Result
+    ): Result
+    many<Result = TArgObj>(
+        receipt: ContractReceipt,
+        fn?: (args: TArgObj) => Result
+    ): Result[]
+
+    waitOne<Result = TArgObj>(
+        source: ContractReceiptSource,
+        fn?: (args: TArgObj) => Result
+    ): Promise<Result>
+    waitMany<Result = TArgObj>(
+        source: ContractReceiptSource,
+        fn?: (args: TArgObj) => Result
+    ): Promise<Result[]>
+
+    waitOneWithReceipt<Result = TArgObj>(
+        source: ContractReceiptSource,
+        fn?: (args: TArgObj) => Result
+    ): Promise<{result: Result; receipt: ContractReceipt}>
+
+    waitManyWithReceipt<Result = TArgObj>(
+        source: ContractReceiptSource,
+        fn?: (args: TArgObj) => Result
+    ): Promise<{result: Result[]; receipt: ContractReceipt}>
+
+    chainOne(
+        source: ContractReceiptSource,
+        fn: (args: TArgObj) => void
+    ): Promise<ContractReceipt>
+    chainMany(
```

Comment:
suggestion: have only one.

The DSL is opinionated, so forcing all users to do things in the same way is encouraged 👍

Reply:
ok